File: backend/app/domains/admin/schemas/public_status.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

PublicComponentState = Literal[
    "operational",
    "degraded",
    "outage",
    "maintenance",
    "unknown",
]


class PublicStatusComponent(BaseModel):
    key: str
    label: str
    status: PublicComponentState
    summary: str
    affected_services: list[str] = Field(default_factory=list)
    updated_at: datetime | None = None


class PublicStatusIncident(BaseModel):
    title: str
    status: str
    severity: str
    kind: Literal["incident", "maintenance"]
    affected_services: list[str] = Field(default_factory=list)
    message: str | None
    started_at: datetime
    resolved_at: datetime | None


class PublicStatusSnapshot(BaseModel):
    generated_at: datetime
    overall_status: PublicComponentState
    headline: str
    summary: str
    components: list[PublicStatusComponent]
    current_incidents: list[PublicStatusIncident]
    scheduled_maintenance: list[PublicStatusIncident]
    recent_history: list[PublicStatusIncident]
    uptime_notice: str

File: frontend/src/lib/api/public-status.ts
import { apiRequest } from "@/lib/api/request";

export type PublicComponentState =
  | "operational"
  | "degraded"
  | "outage"
  | "maintenance"
  | "unknown";

export type PublicStatusComponent = {
  key: string;
  label: string;
  status: PublicComponentState;
  summary: string;
  affected_services: string[];
  updated_at: string | null;
};

export type PublicStatusIncident = {
  title: string;
  status: string;
  severity: string;
  kind: "incident" | "maintenance";
  affected_services: string[];
  message: string | null;
  started_at: string;
  resolved_at: string | null;
};

export type PublicStatusSnapshot = {
  generated_at: string;
  overall_status: PublicComponentState;
  headline: string;
  summary: string;
  components: PublicStatusComponent[];
  current_incidents: PublicStatusIncident[];
  scheduled_maintenance: PublicStatusIncident[];
  recent_history: PublicStatusIncident[];
  uptime_notice: string;
};

export async function getPublicStatusSnapshot(): Promise<PublicStatusSnapshot> {
  return apiRequest<PublicStatusSnapshot>("/status", {
    attachAuth: false,
    attachOrganizationId: false,
    cache: "no-store",
    retry: { maxRetries: 1 },
  });
}
